### ISSUE TITLE
[QUESTION]Exercise 12(Batch Gradient Descent with early stopping for Softmax Regression) from Chapter 4

### ISSUE THREAD

**Issue by Arthurr13** [ageron/handson-ml2#414]
Hi, 

I was going to work on exercise 12(Batch Gradient Descent with early stopping for Softmax Regression) from Chapter 4.

Here is my question: 

1. In [77], Theta = np.random.randn(n_inputs, n_outputs). The theta was created based on the X_train.shape[1] and the number of class in the Iris dataset. 
I am confused why the shape should be based on the X_train.shape[1]. From my understanding, the theta should be the matrix of (length of X_train, len(np.unique(y_train))).Could you give me some suggestion for my understanding?

2. In [77], logits = X_train.dot(Theta). But, it is different with the equation of Softmax score for class k: the theta should theta.T
Could you give me some advice on that?

Many thanks 

Kind regards

**Comment by ageron** [comment]
Hi @Arthurr13 ,

Thanks for your question. You closed it so I guess you found the answer, but I'll answer just in case someone else gets confused about this.

The softmax model will basically learn one linear model per class. If it was for just one class, we would need one bias term (w0) plus one weight per input feature (w1, w2...), just like for linear regression or logistic regression. Since there are 2 input features, that would be a total of 3 parameters (w0, w1, w2). But we need this for each class, and there are 3 classes in the Iris dataset, so that's a total of 3x3=9 parameters. We can store these parameters in a matrix called `Theta`, with one row per parameter, and one column per class. We can initialize this matrix randomly like this:

```python
Theta = np.random.randn(n_inputs, n_outputs)
```

Here, `n_inputs` is the number of input features plus one (for the bias term). That's just the number of columns in `X_train` since it contains one column per feature, plus an extra column full of 1s that we added for the bias term. That's why we defined `n_inputs` using `X_train.shape[1]`.

And `n_outputs` is the number of classes in the training set. One way to find this is to count the number of distinct values in `y_train`, which is `len(np.unique(y_train))`.

I hope this answers your first question.

Now regarding your second question, there's part of the explanation in the note on page 113:

> In Machine Learning, vectors are often represented as _column vectors_, which are 2D arrays with a single column. If **θ** and **x** are column vectors, then the prediction is ŷ=**θ**<sup>⊺</sup> **x**, where **θ**<sup>⊺</sup> is the _transpose_ of **θ** (a row vector instead of a column vector) and **θ**<sup>⊺</sup> **x** is the matrix multiplication of **θ**<sup>⊺</sup> and **x**. It is of course the same prediction, except that it is now represented as a single-cell matrix rather than a scalar value. In this book I will use this notation to avoid switching between dot products and matrix multiplications.

Now note that the dot product is commutative, so ŷ=**θ**<sup>⊺</sup> **x**=**x**<sup>⊺</sup> **θ**.

Okay, now remember that the matrix **X** contains one _row_ per sample, so each row is basically already transposed. That's why we can compute **X** **ϴ** (where **ϴ** is the `Theta` matrix), without having to transpose **X**.

I hope this helps!
